Strange race condition on empy streams

### What were you expecting to happen?

I have the following Vitest-based test in my project:

```ts
import { pipeline } from 'streamx';
import globStream from 'glob-stream';
import toThrough from 'to-through';

describe('globStream', () => {

  it('should concat glob streams', () => new Promise(done => {
    const stream = pipeline(
      globStream('src/*/non-existing.js', { base: 'src' }),
      toThrough(globStream('src/*/non-existing.js', { base: 'src' })),
    );
    stream
      .on('data', () => {})
      .on('end', done);
  }));

});
```

This  test should end without any error. However once in a 10 tries it ends up with timeout (running with vitest). There is a  switched order of `flush` / `end` events in the second vinyl stream. The issue happens more often if the first glob stream actually matches some files.

I am still trying to find the root cause - whether it is in streamx, to-through or gulp-stream (already spent 8 hours getting to this point so I am quite exhausted).

### Terminal output

Successful run:

```console
% npm run -w packages/pipeline test globStream

> <REMOVED-FOR-ISSUE-REPORT>@0.16.1 test
> vitest run globStream


 RUN  v1.6.0 /<REMOVED-FOR-ISSUE-REPORT>

 ✓ src/utils/globStream.spec.ts (1)
   ✓ globStream (1)
     ✓ should concat glob streams

 Test Files  1 passed (1)
      Tests  1 passed (1)
   Start at  23:34:55
   Duration  456ms (transform 79ms, setup 0ms, collect 74ms, tests 11ms, environment 0ms, prepare 151ms)
```

Failed run:

```console
% npm run -w packages/pipeline test globStream

> @<REMOVED-FOR-ISSUE-REPORT>@0.16.1 test
> vitest run globStream


 RUN  v1.6.0 /<REMOVED-FOR-ISSUE-REPORT>

 ❯ src/utils/globStream.spec.ts (1) 5012ms
   ❯ globStream (1) 5010ms
     × should concat glob streams 5010ms

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Tests 1 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 FAIL  src/utils/globStream.spec.ts > globStream > should concat glob streams
Error: Test timed out in 5000ms.
If this is a long-running test, pass a timeout value as the last argument or configure it globally with "testTimeout".
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯

 Test Files  1 failed (1)
      Tests  1 failed (1)
   Start at  23:34:58
   Duration  5.58s (transform 37ms, setup 0ms, collect 48ms, tests 5.01s, environment 0ms, prepare 164ms)

npm ERR! Lifecycle script `test` failed with error: 
npm ERR! Error: command failed 
npm ERR!   in workspace: @<REMOVED-FOR-ISSUE-REPORT>@0.16.1 
npm ERR!   at location: /<REMOVED-FOR-ISSUE-REPORT>
```

### Please provide the following information:

Tested on OSX with NodeJS v20.12.2 and on Kubuntu 24.04 with NodeJS v22.3.0.

--- a/index.js
+++ b/index.js
@@ -261,6 +261,7 @@ function globStream(globs, opt) {
   var stream = new Readable({
     highWaterMark: ourOpt.highWaterMark,
     read: read,
+    open: open,
     predestroy: predestroy,
   });
 
@@ -277,19 +278,23 @@ function globStream(globs, opt) {
   walker.on('path', onPath);
   walker.once('end', onEnd);
   walker.once('error', onError);
-  ourGlobs.forEach(function (glob) {
-    if (isGlob(glob)) {
-      // We only want to walk the glob-parent directories of any positive glob
-      // to reduce the amount of files have to check.
-      if (isPositiveGlob(glob)) {
-        var base = globParent(glob);
-        walker.walk(base);
+
+  function open(cb) {
+    ourGlobs.forEach(function (glob) {
+      if (isGlob(glob)) {
+        // We only want to walk the glob-parent directories of any positive glob
+        // to reduce the amount of files have to check.
+        if (isPositiveGlob(glob)) {
+          var base = globParent(glob);
+          walker.walk(base);
+        }
+      } else {
+        // If the strig is not a glob, we just check for the existence of it.
+        walker.exists(glob);
       }
-    } else {
-      // If the strig is not a glob, we just check for the existence of it.
-      walker.exists(glob);
-    }
-  });
+    });
+    cb();
+  }
 
   function read(cb) {
     walker.resume();

--- a/index.js
+++ b/index.js
@@ -289,7 +289,7 @@ function globStream(globs, opt) {
           walker.walk(base);
         }
       } else {
-        // If the strig is not a glob, we just check for the existence of it.
+        // If the string is not a glob, we just check for the existence of it.
         walker.exists(glob);
       }
     });

--- a/test/index.js
+++ b/test/index.js
@@ -887,6 +887,22 @@ function suite(moduleName) {
         done
       );
     });
+
+    it('does not end prematurely', function (done) {
+      var gs = globStream(['./non-existent-file'], { cwd: dir, allowEmpty: true });
+
+      function setup() {
+        stream.pipeline(
+          [
+            gs,
+            concat(),
+          ],
+          done
+        );
+      }
+
+      setTimeout(setup, 10);
+    });
   });
 
   describe('options', function () {

--- a/test/index.js
+++ b/test/index.js
@@ -889,19 +889,28 @@ function suite(moduleName) {
     });
 
     it('does not end prematurely', function (done) {
-      var gs = globStream(['./non-existent-file'], { cwd: dir, allowEmpty: true });
+      var gs = globStream(['./fixtures/*.js'], { cwd: dir });
 
-      function setup() {
+      function delayed() {
         stream.pipeline(
           [
             gs,
+            new stream.Transform({
+              objectMode: true,
+              transform: function(data, enc, cb) {
+                if (typeof enc === 'function') {
+                  cb = enc;
+                }
+                cb(null, data);
+              }
+            }),
             concat(),
           ],
           done
         );
       }
 
-      setTimeout(setup, 10);
+      setTimeout(delayed, 10);
     });
   });
 

